
<!DOCTYPE html>
<html lang="zh-CN ">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>ITthreeの博客 || 异常</title>
    <meta name="author" content="ITthree">
    <meta name="description" content="ITthreeの博客,我思故我在 ">
    <meta name="keywords" content=" ">
    <link rel="icon" href="/../images/icon.jpg">
    <link rel="stylesheet" href="/css/antd.min.css">
    
    <link rel="stylesheet" href="/css/night-theme.css">
    
    <script src="/js/vue.js"></script>
    <script src="/js/antd.min.js"></script>
<meta name="generator" content="Hexo 5.0.1"></head>

<body>

    <div id="loading"
        style="height: 100vh; width: 100%; position: fixed;display: flex;z-index: 200; justify-content: space-between;">
        <div id="loadleft" style="width: 50%;background-color: #ffffff;transition: width 0.6s ease-out;"></div>
        <div id="loadright" style="width: 50%;background-color: #ffffff;transition: width 0.6s ease-out;"></div>
        <div
            style="position: fixed; height: 100vh; width: 100%;display: flex;justify-content: center;align-items: center;">
            <div id="loadcontent"
                style="width:400px;height:400px;padding:50px;border-radius:50%;display:flex;justify-content:center;align-items:center;border:solid 10px#a3ddfb; text-align:center;opacity:1;transition:opacity 0.3s ease-out;">
                <div>
                    <h2>LOADING...</h2>
                    <p>加载过慢请开启缓存(浏览器默认开启)</p>
                    <div>
                        <img src="/dancingkitty.gif" alt="loading">
                    </div>
                </div>
            </div>
        </div>
    </div>

    <div id="layout">
        <transition name="into">
            <div v-show="show_page" style="display: none;">
                <div id="menu_show">
                     
<nav id="menu">
    <div class="desktop-menu">
        <a href="/">
            <span class="title">ITthreeの博客</span>
        </a>
        
        <a href="/">
            <span>
                <a-icon type="home" theme="filled" />
            </span>
            <span>home</span>
        </a>
        
        <a href="/about">
            <span>
                <a-icon type="idcard" theme="filled" />
            </span>
            <span>about</span>
        </a>
        
        <a href="/archives">
            <span>
                <a-icon type="folder" theme="filled" />
            </span>
            <span>archives</span>
        </a>
        
        <a href="/categories/ubuntu">
            <span>
                <a-icon type="book" theme="filled" />
            </span>
            <span>categories</span>
        </a>
        
        <a href="/tags/shell">
            <span>
                <a-icon type="tags" theme="filled" />
            </span>
            <span>tags</span>
        </a>
        
    </div>

    <div :class="'phone-menu ' + menu_show" id="phone-menu">
        <div :class="'title'" @click="menu_show=!menu_show">
            <span style="margin-right: 10px;">
                <a-icon type="appstore" theme="filled" />
            </span>
            <span>ITthreeの博客</span>
        </div>
        <div class="items" v-show="menu_show">
            
            <a href="/">
                <div class="item">
                    <div style="min-width:20px; max-width:50px; width: 10%">
                        <a-icon type="home" theme="filled" />
                    </div>
                    <div style="min-width:100px;max-width: 150%;width: 20%;">home</div>
                </div>
            </a>
            
            <a href="/about">
                <div class="item">
                    <div style="min-width:20px; max-width:50px; width: 10%">
                        <a-icon type="idcard" theme="filled" />
                    </div>
                    <div style="min-width:100px;max-width: 150%;width: 20%;">about</div>
                </div>
            </a>
            
            <a href="/archives">
                <div class="item">
                    <div style="min-width:20px; max-width:50px; width: 10%">
                        <a-icon type="folder" theme="filled" />
                    </div>
                    <div style="min-width:100px;max-width: 150%;width: 20%;">archives</div>
                </div>
            </a>
            
            <a href="/categories/ubuntu">
                <div class="item">
                    <div style="min-width:20px; max-width:50px; width: 10%">
                        <a-icon type="book" theme="filled" />
                    </div>
                    <div style="min-width:100px;max-width: 150%;width: 20%;">categories</div>
                </div>
            </a>
            
            <a href="/tags/shell">
                <div class="item">
                    <div style="min-width:20px; max-width:50px; width: 10%">
                        <a-icon type="tags" theme="filled" />
                    </div>
                    <div style="min-width:100px;max-width: 150%;width: 20%;">tags</div>
                </div>
            </a>
            
        </div>
        <div class="curtain" v-show="menu_show"></div>
    </div>

</nav>
                </div>

                <div id="main">
                     
<link rel="stylesheet" href="/css/post-body.css">
<div class="article">
    <div>
        <h1>异常 </h1>
    </div>
    <div class="info">
        <span class="date">
            <span class="icon">
                <a-icon type="calendar" theme="filled" />
            </span>
            2020/10/16
        </span>

        
        <span class="category">
            <a href="/categories/Java">
                <span class="icon">
                    <a-icon type="book" theme="filled" />
                </span>
                Java
            </a>
        </span>
        

        

        <span class="tags">
            <span class="icon">
                <a-icon type="tags" theme="filled" />
            </span>
            
            <span class="tag">
                
                <a href="/tags/Java基础" style=color:#607d8b>
                    Java基础
                </a>
            </span>
            
        </span>
        
    </div>

    <div class="content" v-pre>
        <p><img src="http://img.lbkzw.cn/img/20200821102446.png" alt="img"></p>
<ul>
<li><p>RuntimeException(运行时异常)会自动被 java 虚拟机抛出，所以不必在异常说明中把它们列出来,也不需要在异常说明中声明方法将抛出 RuntimeException 类型的异常</p>
</li>
<li><p>RuntimeException 也被称为“不受检查异常”，将被自动捕获,如果 没有被捕获而直达 main()，那么在程序退出前将调用异常的 printStackTrace() 方法。(返回栈轨迹信息)</p>
</li>
<li><p>要么<strong>声明</strong>异常要么<strong>处理</strong>异常(只有在你知道如何处理的情况下才捕获异常)。</p>
<pre><code class="java">package cn.day.a2;

// exceptions/LostMessage.java
// How an exception can be lost
class VeryImportantException extends Exception &amp;#123;
    @Override
    public String toString() &amp;#123;
        return &quot;A very important exception!&quot;;
    &amp;#125;
&amp;#125;
class HoHumException extends Exception &amp;#123;
    @Override
    public String toString() &amp;#123;
        return &quot;A trivial exception&quot;;
    &amp;#125;
&amp;#125;
public class LostMessage &amp;#123;
    void f() throws VeryImportantException &amp;#123;
        throw new VeryImportantException();
    &amp;#125;
    void dispose() throws HoHumException &amp;#123;
        throw new HoHumException();
    &amp;#125;
    /*声明异常，而不做处理（交给调用的人处理）*/
    public static void main(String[] args) throws VeryImportantException &amp;#123;//仅仅声明异常

            LostMessage lm = new LostMessage();
                lm.f();

            &amp;#125;

    &amp;#125;
    /*捕获并处理异常*/
//    public static void main(String[] args)  &amp;#123;
//        try &amp;#123;
//            LostMessage lm = new LostMessage();
//            try &amp;#123;
//                lm.f();
//            &amp;#125; finally &amp;#123;
//                lm.dispose();
//            &amp;#125;
//        &amp;#125; catch(VeryImportantException | HoHumException e) &amp;#123;
//            System.out.println(e);
//        &amp;#125;
//        
//    &amp;#125;
//
//&amp;#125;

</code></pre>
</li>
</ul>
<pre><code>


* 可以通过**继承**异常基类来**自定义异常**

```java
class SimpleException extends Exception &amp;#123;&amp;#125;

public class InheritingExceptions &amp;#123;
    public void f() throws SimpleException &amp;#123;
        System.out.println(
                &quot;Throw SimpleException from f()&quot;);
        throw new SimpleException();
    &amp;#125;
    public static void main(String[] args) &amp;#123;
        InheritingExceptions sed =
                new InheritingExceptions();
        try &amp;#123;
            sed.f();
        &amp;#125; catch(SimpleException e) &amp;#123;
            System.out.println(&quot;Caught it!&quot;);
        &amp;#125;
    &amp;#125;
&amp;#125;
</code></pre>
<h1 id="声明异常"><a href="#声明异常" class="headerlink" title="声明异常"></a>声明异常</h1><h2 id="声明多个异常"><a href="#声明多个异常" class="headerlink" title="声明多个异常"></a>声明多个异常</h2><pre><code class="java">void f() throws TooBig, TooSmall, DivZero &amp;#123; // 方法体...&amp;#125;</code></pre>
<h2 id="声明所有异常"><a href="#声明所有异常" class="headerlink" title="声明所有异常"></a>声明所有异常</h2><pre><code class="java">catch(Exception e) &amp;#123;
    System.out.println(&quot;Caught an exception&quot;);
&amp;#125;</code></pre>
<h2 id="声明子异常捕获父异常"><a href="#声明子异常捕获父异常" class="headerlink" title="声明子异常捕获父异常"></a>声明子异常捕获父异常</h2><blockquote>
<p>java7之后就可以通过编译了</p>
</blockquote>
<pre><code class="java">class BaseException extends Exception &amp;#123;&amp;#125;
class DerivedException extends BaseException &amp;#123;&amp;#125;

public class PreciseRethrow &amp;#123;
    void catcher() throws DerivedException &amp;#123;//声明子异常
        try &amp;#123;
            throw new DerivedException();
        &amp;#125; catch(BaseException e) &amp;#123;//捕获父异常
            throw e;
        &amp;#125;
    &amp;#125;
&amp;#125;
</code></pre>
<h1 id="捕获异常"><a href="#捕获异常" class="headerlink" title="捕获异常"></a>捕获异常</h1><h2 id="捕获多个异常"><a href="#捕获多个异常" class="headerlink" title="捕获多个异常"></a>捕获多个异常</h2><pre><code class="java">// exceptions/SameHandler.java
class EBase1 extends Exception &amp;#123;&amp;#125;
class Except1 extends EBase1 &amp;#123;&amp;#125;
class EBase2 extends Exception &amp;#123;&amp;#125;
class Except2 extends EBase2 &amp;#123;&amp;#125;
class EBase3 extends Exception &amp;#123;&amp;#125;
class Except3 extends EBase3 &amp;#123;&amp;#125;
class EBase4 extends Exception &amp;#123;&amp;#125;
class Except4 extends EBase4 &amp;#123;&amp;#125;

public class SameHandler &amp;#123;
    void x() throws Except1, Except2, Except3, Except4 &amp;#123;&amp;#125;
    void process() &amp;#123;&amp;#125;
    void f() &amp;#123;
        try &amp;#123;
            x();
            /*以下是分别捕获并处理异常*/
        &amp;#125; catch(Except1 e) &amp;#123;
            process();
        &amp;#125; catch(Except2 e) &amp;#123;
            process();
        &amp;#125; catch(Except3 e) &amp;#123;
            process();
        &amp;#125; catch(Except4 e) &amp;#123;
            process();
        &amp;#125;
    &amp;#125;
&amp;#125;</code></pre>
<h3 id="多重捕获"><a href="#多重捕获" class="headerlink" title="多重捕获"></a>多重捕获</h3><ul>
<li>java7之后即使不同类型的异常也可通过**|**来连接捕获，进行异常处理</li>
</ul>
<pre><code>// exceptions/MultiCatch.java
public class MultiCatch &amp;#123;
    void x() throws Except1, Except2, Except3, Except4 &amp;#123;&amp;#125;
    void process() &amp;#123;&amp;#125;
    void f() &amp;#123;
        try &amp;#123;
            x();

        &amp;#125; catch(Except1 | Except2 | Except3 | Except4 e) &amp;#123;
            process();
        &amp;#125;
    &amp;#125;
&amp;#125;
</code></pre>
<h2 id="重新抛出异常"><a href="#重新抛出异常" class="headerlink" title="重新抛出异常"></a>重新抛出异常</h2><ul>
<li>重抛异常会把异常抛给上一级环境中的异常处理程序，同一个 try 块的后续 catch 子句将被忽略。此外，异常对象的所有信息都得以保持，所以高一级环境中捕获此异常的处理程序可以从这个异常对象中得到所有信息。</li>
</ul>
<pre><code class="java">catch(Exception e) &amp;#123;
    System.out.println(&quot;An exception was thrown&quot;);
    throw e;
&amp;#125;</code></pre>
<ul>
<li>当再次抛出异常时,printStackTrace() 方法显示的将是原来异常抛出点调用栈的信息，而并非重新抛出点的信息。可以调用 filInStackTrace() 方法更新异常抛出点调用栈的信息（这将返回一个 Throwable 对象）。</li>
</ul>
<pre><code class="java">package cn.day.a1;

public class Rethrowing &amp;#123;
    public static void f() throws Exception &amp;#123;
        System.out.println(
                &quot;originating the exception in f()&quot;);
        throw new Exception(&quot;thrown from f()&quot;);
    &amp;#125;
    public static void g() throws Exception &amp;#123;
        try &amp;#123;
            f();
        &amp;#125; catch(Exception e) &amp;#123;
            System.out.println(
                    &quot;Inside g(), e.printStackTrace()&quot;);
            e.printStackTrace(System.out);
            throw e;//再次抛出后没有更新
        &amp;#125;
    &amp;#125;
    public static void h() throws Exception &amp;#123;
        try &amp;#123;
            f();
        &amp;#125; catch(Exception e) &amp;#123;
            System.out.println(
                    &quot;Inside h(), e.printStackTrace()&quot;);
            e.printStackTrace(System.out);
            throw (Exception)e.fillInStackTrace();//抛出并更新
        &amp;#125;
    &amp;#125;
    public static void main(String[] args) &amp;#123;
        try &amp;#123;
            g();//抛出点的调用栈信息一致
        &amp;#125; catch(Exception e) &amp;#123;
            System.out.println(&quot;main: printStackTrace()&quot;);
            e.printStackTrace(System.out);

        &amp;#125;
        try &amp;#123;
            h();//抛出点的调用栈信息是更新前后的
        &amp;#125; catch(Exception e) &amp;#123;
            System.out.println(&quot;main: printStackTrace()&quot;);
            e.printStackTrace(System.out);
        &amp;#125;
    &amp;#125;
&amp;#125;</code></pre>
<ul>
<li>在捕获异常之后抛出另一种异常,效果类似于使用 filInStackTrace()，有关原异常抛出点调用栈会丢失</li>
</ul>
<pre><code class="java">// exceptions/RethrowNew.java
// Rethrow a different object from the one you caught
class OneException extends Exception &amp;#123;
    OneException(String s) &amp;#123; super(s); &amp;#125;
&amp;#125;
class TwoException extends Exception &amp;#123;
    TwoException(String s) &amp;#123; super(s); &amp;#125;
&amp;#125;
public class RethrowNew &amp;#123;
    public static void f() throws OneException &amp;#123;
        System.out.println(
                &quot;originating the exception in f()&quot;);
        throw new OneException(&quot;thrown from f()&quot;);
    &amp;#125;
    public static void main(String[] args) &amp;#123;
        try &amp;#123;
            try &amp;#123;
                f();
            &amp;#125; catch(OneException e) &amp;#123;
                System.out.println(
                        &quot;Caught in inner try, e.printStackTrace()&quot;);
                e.printStackTrace(System.out);
                throw new TwoException(&quot;from inner try&quot;);//抛出不同异常
            &amp;#125;
        &amp;#125; catch(TwoException e) &amp;#123;
            System.out.println(
                    &quot;Caught in outer try, e.printStackTrace()&quot;);
            e.printStackTrace(System.out);//此异常仅知道自己来自 main()，而对 f() 一无所知。
        &amp;#125;
    &amp;#125;
&amp;#125;
</code></pre>
<h1 id="异常链"><a href="#异常链" class="headerlink" title="异常链"></a>异常链</h1><p>捕获一个异常并抛出另一个异常，同时想保持已捕获异常的信息。</p>
<h1 id="finally"><a href="#finally" class="headerlink" title="finally"></a>finally</h1><ul>
<li><p>异常有无发生，均执行的操作（抛出异常时紧跟异常处理语句后执行）</p>
<pre><code class="java">// exceptions/MultipleReturns.java
public class MultipleReturns &amp;#123;
    public static void f(int i) &amp;#123;
        System.out.println(
                &quot;Initialization that requires cleanup&quot;);
        try &amp;#123;
            System.out.println(&quot;Point 1&quot;);
            if(i == 1) return;//返回，中断下面操作，但finally语句仍会执行。
            System.out.println(&quot;Point 2&quot;);
            if(i == 2) return;
            System.out.println(&quot;Point 3&quot;);
            if(i == 3) return;
            System.out.println(&quot;End&quot;);
            return;
        &amp;#125; finally &amp;#123;
            System.out.println(&quot;Performing cleanup&quot;);
        &amp;#125;
    &amp;#125;
    public static void main(String[] args) &amp;#123;
        for(int i = 1; i &lt;= 4; i++)
            f(i);
    &amp;#125;
&amp;#125;</code></pre>
</li>
<li><p>在任何finally语句中使用return，发生(抛出)的异常将静默。</p>
<pre><code class="javascript">public class ExceptionSilencer &amp;#123;
    public static void main(String[] args) &amp;#123;
        try &amp;#123;
            throw new RuntimeException();
        &amp;#125; finally &amp;#123;
            // Using &#39;return&#39; inside the finally block
            // will silence any thrown exception.
            return;
        &amp;#125;
    &amp;#125;
&amp;#125;</code></pre>
</li>
<li><p>try中与finally中发生同时（抛出）异常，try中的异常将不被显示(设计缺陷)</p>
<pre><code class="java">package cn.day.a2;

// exceptions/LostMessage.java
// How an exception can be lost
class VeryImportantException extends Exception &amp;#123;
    @Override
    public String toString() &amp;#123;
        return &quot;A very important exception!&quot;;
    &amp;#125;
&amp;#125;
class HoHumException extends Exception &amp;#123;
    @Override
    public String toString() &amp;#123;
        return &quot;A trivial exception&quot;;
    &amp;#125;
&amp;#125;
public class LostMessage &amp;#123;
    void f() throws VeryImportantException &amp;#123;
        throw new VeryImportantException();
    &amp;#125;
    void dispose() throws HoHumException &amp;#123;
        throw new HoHumException();
    &amp;#125;
    public static void main(String[] args) &amp;#123;
        try &amp;#123;
            LostMessage lm = new LostMessage();
            try &amp;#123;
                lm.f();//由于finally也抛出异常，故此异常不被显示
            &amp;#125; finally &amp;#123;
                lm.dispose();
            &amp;#125;
        &amp;#125; catch(VeryImportantException | HoHumException e) &amp;#123;
            System.out.println(e);
        &amp;#125;
    &amp;#125;
&amp;#125;
</code></pre>
</li>
</ul>
<h1 id="异常限制"><a href="#异常限制" class="headerlink" title="异常限制"></a>异常限制</h1><ul>
<li>一个类可不声明其超类或实现接口中的异常</li>
<li>一个类不能声明与其超类中声明的异常无直接继承关系的异常</li>
<li>异常限制对构造器不起作用。</li>
</ul>
<pre><code class="java">package cn.day.a2;

/**重写的方法只能抛出在其基类版本中指定的异常，或从基类异常派生的异常**/
/*定义异常*/
class BaseballException extends Exception &amp;#123;&amp;#125;
class Foul extends BaseballException &amp;#123;&amp;#125;
class Strike extends BaseballException &amp;#123;&amp;#125;

/*声明异常*/
abstract class Inning &amp;#123;
    Inning() throws BaseballException &amp;#123;&amp;#125;
    public void event() throws BaseballException &amp;#123;
    &amp;#125;
    public abstract void atBat() throws Strike, Foul;
    public void walk() &amp;#123;&amp;#125; 
&amp;#125;
/*定义异常*/
class StormException extends Exception &amp;#123;&amp;#125;
class RainedOut extends StormException &amp;#123;&amp;#125;
class PopFoul extends Foul &amp;#123;&amp;#125;

/*一个接口中声明两个异常*/
interface Storm &amp;#123;
    void event() throws RainedOut;
    void rainHard() throws RainedOut;
&amp;#125;

/*继承上面含有声明异常的类并实现含有声明异常的接口*/
public class StormyInning extends Inning implements Storm &amp;#123;

    public StormyInning()
            throws RainedOut, BaseballException &amp;#123;&amp;#125;//声明这个构造器可能发生的异常（来自接口与父类）
    public StormyInning(String s)
  throws BaseballException &amp;#123;&amp;#125;

    @Override
    public void rainHard() throws RainedOut &amp;#123;&amp;#125;
//声明了实现的接口异常
    @Override
    public void event() &amp;#123;&amp;#125;
    // 没有声明异常哦
    @Override
    public void atBat() throws PopFoul &amp;#123;&amp;#125;//覆盖方法可以声明继承了其超类所声明的异常的子异常
//    public void atBat() throws RainedOut &amp;#123;&amp;#125;//其接口有，但其超类未声明此异常，故不可这样做
//    public void atBat() throws BaseballException &amp;#123;&amp;#125;覆盖方法不能声明其超类中未声明且与其超类中声明的异常无直接继承关系的异常
    public static void main(String[] args) &amp;#123;
        try &amp;#123;
            StormyInning si = new StormyInning();
            si.atBat();
        &amp;#125; catch(PopFoul e) &amp;#123;
            System.out.println(&quot;Pop foul&quot;);
        &amp;#125; catch(RainedOut e) &amp;#123;
            System.out.println(&quot;Rained out&quot;);
        &amp;#125; catch(BaseballException e) &amp;#123;
            System.out.println(&quot;Generic baseball exception&quot;);
        &amp;#125;
        try &amp;#123;
            Inning i = new StormyInning();
            i.atBat();
        &amp;#125; catch(Strike e) &amp;#123;
            System.out.println(&quot;Strike&quot;);
        &amp;#125; catch(Foul e) &amp;#123;
            System.out.println(&quot;Foul&quot;);
        &amp;#125; catch(RainedOut e) &amp;#123;
            System.out.println(&quot;Rained out&quot;);
        &amp;#125; catch(BaseballException e) &amp;#123;
            System.out.println(&quot;Generic baseball exception&quot;);
        &amp;#125;
    &amp;#125;
&amp;#125;</code></pre>
<h1 id="异常匹配"><a href="#异常匹配" class="headerlink" title="异常匹配"></a>异常匹配</h1><ul>
<li><p>异常处理，采用“就近原则”：异常子类可作为捕获参数用于捕获其父类类型的异常。</p>
<pre><code class="java">// exceptions/Human.java
// Catching exception hierarchies
class Annoyance extends Exception &amp;#123;&amp;#125;
class Sneeze extends Annoyance &amp;#123;&amp;#125;
public class Human &amp;#123;
    public static void main(String[] args) &amp;#123;
        // Catch the exact type:
        try &amp;#123;
            throw new Sneeze();
        &amp;#125; catch(Sneeze s) &amp;#123;
            System.out.println(&quot;Caught Sneeze&quot;);
        &amp;#125; catch(Annoyance a) &amp;#123;
            System.out.println(&quot;Caught Annoyance&quot;);
        &amp;#125;

        try &amp;#123;
            throw new Sneeze();
        &amp;#125; catch(Annoyance a) &amp;#123;//可捕获Annoyance父类Sneeze的异常
            System.out.println(&quot;Caught Annoyance&quot;);
        &amp;#125;
         catch(Sneeze a) &amp;#123;//将永远无法捕获并处理异常，所以会报错。
            System.out.println(&quot;Caught Annoyance&quot;);
        &amp;#125;
    &amp;#125;
&amp;#125;</code></pre>
</li>
</ul>
<h1 id="try-with-resources"><a href="#try-with-resources" class="headerlink" title="try-with-resources"></a>try-with-resources</h1><ul>
<li>确保有无异常时打开的文件在结束之后均关闭</li>
</ul>
<pre><code class="java">static String readFirstLineFromFile(String path) throws IOException &amp;#123;
      try (BufferedReader br = new BufferedReader(new FileReader(path))) &amp;#123;//即使发生异常（）内的资源也会关闭
        return br.readLine();
      &amp;#125;
    &amp;#125;

/*try finally*/
static String readFirstLineFromFileWithFinallyBlock(String path) throws IOException &amp;#123;
      BufferedReader br = new BufferedReader(new FileReader(path));
      try &amp;#123;
        return br.readLine();
      &amp;#125; finally &amp;#123;
        if (br != null) br.close();//如果close出现异常，则文件无法正常关闭
      &amp;#125;
    &amp;#125;</code></pre>

    </div>

    
</div>
                     
<footer id="footer">
    <div class="footer-wrap">
        <div>
            © 2020 - 2021 ITthreeの博客
            <span class="footer-icon">
                <a-icon type="flag" theme="filled" /></span>
            @ITthree
        </div>
        <div></div>
        <div>Based on the <a target="_blank" rel="noopener" href="https://hexo.io/">Hexo Engine</a> & <a
                target="_blank" rel="noopener" href="https://github.com/korilin/hexo-theme-particle">Particle Theme</a></div>
        
    </div>

</footer>

<script src="/js/highlight.min.js"></script>
<script src="/js/particle.js"></script>
                </div>
            </div>
        </transition>
    </div>

    <script>
    new Vue({
        el: "#layout",
        data: {
            show_page: false,
            onload_menu: false,
            menu_show: false,
            card_top: 100
        },
        created: function () {
            var that = this
            window.onload = function () {
                that.show_page = true
                document.getElementById("loadcontent").style.opacity = 0
                setTimeout(function () {
                    document.getElementById("loadleft").style.width = 0
                    document.getElementById("loadright").style.width = 0
                }, 300)
                setTimeout(function () {
                    document.getElementById("loading").style = "display:none"
                }, 600)
            }
        },
        mounted: function () {
            var that = this
            window.addEventListener('scroll', function (e) {
                that.menu_show = false
            })
        },
        methods: {
            home_click: function () {
                window.scrollTo({
                    top: window.innerHeight - 80,
                    behavior: "smooth",
                });
            }
        }
    })
</script>

</body>

</html>